service/qos/service_level_controller.cc

```diff
@@ -404,6 +426,105 @@ future<> service_level_controller::do_add_service_level(sstring name, service_le
     return make_ready_future();
 }
 
+bool service_level_controller::is_v2() const {
+    return _sl_data_accessor && _sl_data_accessor->is_v2();
+}
+
+void service_level_controller::upgrade_to_v2(cql3::query_processor& qp, service::raft_group0_client& group0_client) {
+    if (!_sl_data_accessor) {
+        return;
+    }
+
+    auto v2_data_accessor = _sl_data_accessor->upgrade_to_v2(qp, group0_client);
+    if (v2_data_accessor) {
+        _sl_data_accessor = v2_data_accessor;
+    }
+}
+
+future<> service_level_controller::migrate_to_v2(size_t nodes_count, db::system_keyspace& sys_ks, cql3::query_processor& qp, service::raft_group0_client& group0_client, abort_source& as) {
+    //TODO:
+    //Now we trust the administrator to not make changes to service levels during the migration.
+    //Ideally, during the migration we should set migration data accessor(on all nodes, on all shards) that allows to read but forbids writes
+    using namespace std::chrono_literals;
+
+    auto migration_status = co_await sys_ks.get_service_levels_migration_status();
+    if (migration_status == db::system_keyspace::SERVICE_LEVELS_MIGRATED_VALUE) {
+        // service levels were already migrated to v2 table
+        co_return;
+    }
+
+    auto schema = qp.db().find_schema(db::system_distributed_keyspace::NAME, db::system_distributed_keyspace::SERVICE_LEVELS);
+
+    const auto t = 5min;
+    const timeout_config tc{t, t, t, t, t, t, t};
+    service::client_state cs(::service::client_state::internal_tag{}, tc);
+    service::query_state qs(cs, empty_service_permit());
+    
+    // `system_distributed` keyspace has RF=3 and we need to scan it with CL=ALL
+    // To support migration on cluster with 1 or 2 nodes, set appropriate CL
+    auto cl = db::consistency_level::ALL;
+    if (nodes_count == 1) {
```

Comment:
this is odd, really ALL doesn't work for 1-2 node clusters?

Reply:
yes, it looks like consistency ALL requires as many alive replicas as repliaction factor of the ks (I don't know if it's expected or a bug?) and `system_distributed` as RF=3 initially.

Reply:
ok